klippy/webhooks.py

```diff
@@ -280,3 +280,136 @@ def get_status(self, eventtime=0.):
+        self.available_objects = {}
+        objs = self.printer.lookup_objects()
+        status_objs = {n: o for n, o in objs if hasattr(o, "get_status")}
+        for name, obj in status_objs.items():
+            attrs = obj.get_status(eventtime)
+            self.available_objects[name] = attrs.keys()
+        self.ready = True
+
+    def _handle_restart(self, eventtime):
+        self.ready = False
+        self.reactor.update_timer(self.subscription_timer, self.reactor.NEVER)
+
+    def _batch_subscription_handler(self, eventtime):
+        status = self._process_status_request(self.subscriptions)
+        self.webhooks.call_remote_method(
+            "process_status_update", status=status)
+        return eventtime + SUBSCRIPTION_REFRESH_TIME
+
+    def _process_status_request(self, requested_objects):
+        result = {}
+        if self.ready:
+            for name, req_items in requested_objects.items():
+                obj = self.printer.lookup_object(name, None)
+                if obj is not None and name in self.available_objects:
+                    status = obj.get_status(self.reactor.monotonic())
+                    if not req_items:
+                        # return all items excluding callables
+                        result[name] = {k: v for k, v in status.items()
+                                        if not callable(v)}
+                    else:
+                        # return requested items excluding callables
+                        result[name] = {k: v for k, v in status.items()
+                                        if k in req_items and not callable(v)}
+        else:
+            result = {"status": "Klippy Not Ready"}
+        return result
+
+    def _handle_object_request(self, web_request):
+        if web_request.get_method() != 'GET':
+            raise web_request.error("Invalid Request Method")
+        web_request.send(dict(self.available_objects))
+
+    def _handle_status_request(self, web_request):
+        if web_request.get_method() != 'GET':
+            raise web_request.error("Invalid Request Method")
+        args = web_request.get_args()
+        result = self._process_status_request(args)
+        web_request.send(result)
+
+    def _handle_subscription_request(self, web_request):
+        method = web_request.get_method()
+        if method == 'POST':
+            # add a subscription
+            args = web_request.get_args()
+            if args:
+                self.add_subscripton(args)
+            else:
+                raise web_request.error("Invalid argument")
+        else:
+            # get subscription info
+            result = dict(self.subscriptions)
+            web_request.send(result)
+
+    def add_subscripton(self, new_sub):
+        if not new_sub:
+            return
+        for obj_name, req_items in new_sub.items():
+            if obj_name not in self.available_objects or \
+                    obj_name.startswith("gcode_macro"):
```

Comment:
Out of curiosity, why is gcode_macro excluded?

Reply:
My original thought was that gcode macros are user defined and can't be relied upon for consistent behavior.   Perhaps this is a bit too much hand holding.  Allowing them in could allow for some interesting "custom" clients, so I'll remove the check.